File: scripts/sci_to_mermaid.ts
import * as fs from 'fs';

function sciToMermaid(sciPath: string): string | undefined {
    if (!fs.existsSync(sciPath)) {
        console.error(`Error: ${sciPath} not found.`);
        return undefined;
    }

    const xmlData = fs.readFileSync(sciPath, 'utf-8');
    
    // 1. Isolate the InternalAllocatedLinkIds section
    const linkIdsMatch = xmlData.match(/<InternalAllocatedLinkIds[^>]*>(.*?)<\/InternalAllocatedLinkIds>/);
    if (!linkIdsMatch) {
        console.error("Error: Could not find InternalAllocatedLinkIds section.");
        return undefined;
    }
    const linkIdsSection = linkIdsMatch[1];

    // 2. Extract Key/Value pairs from this section only
    const links: [string, string][] = [];
    const entryRegex = /<a:KeyValueOfstringstring>.*?<a:Key[^>]*>(.*?)<\/a:Key>.*?<a:Value[^>]*>(.*?)<\/a:Value>.*?<\/a:KeyValueOfstringstring>/g;

    let match;
    while ((match = entryRegex.exec(linkIdsSection)) !== null) {
        let key = match[1];
        let val = match[2];
        
        // Decode XML entities
        key = key.replace(/&gt;/g, '>').replace(/&lt;/g, '<').replace(/&amp;/g, '&');
        val = val.replace(/&gt;/g, '>').replace(/&lt;/g, '<').replace(/&amp;/g, '&');
        
        links.push([key, val]);
    }

    if (links.length === 0) {
        console.error("Error: No transitions found in the SCI file.");
        return "stateDiagram-v2\n    Note over Error: No transitions found";
    }

    const machines: Record<string, string[]> = {};
    const mermaid: string[] = ["stateDiagram-v2"];

    for (const [key, val] of links) {
        // Format: PImplementation.Machine.SourceState->PImplementation.Machine.TargetState(id)
        if (key.includes("->")) {
            const parts = key.split("->");
            const sourceFull = parts[0];
            const targetFull = parts[1].split("(")[0];
            
            const sourceParts = sourceFull.split(".");
            const targetParts = targetFull.split(".");
            
            if (sourceParts.length >= 3 && targetParts.length >= 3) {
                const machineName = sourceParts[1];
                const sourceState = sourceParts[2];
                const targetState = targetParts[2];
                
                // Only include if it's a "normal" transition (ignoring internal IDs)
                if (val !== "0" && val !== "1") {
                    if (!machines[machineName]) {
                        machines[machineName] = [];
                    }
                    machines[machineName].push(`        ${sourceState} --> ${targetState}: ${val}`);
                }
            }
        }
    }

    for (const machine in machines) {
        mermaid.push(`    state ${machine} {`);
        const uniqueTransitions = Array.from(new Set(machines[machine])).sort();
        mermaid.push(...uniqueTransitions);
        mermaid.push("    }");
    }

    return mermaid.join("\n");
}

const args = process.argv.slice(2);
if (args.length < 1) {
    console.log("Usage: ts-node sci_to_mermaid.ts <file.sci>");
    process.exit(1);
}

const result = sciToMermaid(args[0]);
if (result) {
    console.log(result);
}

File: scripts/sci_to_mermaid.py
import sys
import os
import xml.etree.ElementTree as ET

def sci_to_mermaid(sci_path):
    if not os.path.exists(sci_path):
        print(f"Error: {sci_path} not found.")
        return

    try:
        tree = ET.parse(sci_path)
        root = tree.getroot()
    except Exception as e:
        print(f"Error parsing XML: {e}")
        return

    # Use a namespace-agnostic search for KeyValueOfstringstring
    links = []
    for elem in root.iter():
        if 'KeyValueOfstringstring' in elem.tag:
            key_elem = None
            val_elem = None
            for child in elem:
                if 'Key' in child.tag:
                    key_elem = child
                if 'Value' in child.tag:
                    val_elem = child
            
            if key_elem is not None and key_elem.text and val_elem is not None and val_elem.text:
                links.append((key_elem.text, val_elem.text))
    
    if not links:
        print("Error: No transitions found in the SCI file.")
        return "stateDiagram-v2\n    Note over Error: No transitions found"

    machines = {}
    mermaid = ["stateDiagram-v2"]

    for key, val in links:
        # Format: PImplementation.Machine.SourceState->PImplementation.Machine.TargetState(id)
        if "->" in key:
            parts = key.split("->")
            source_full = parts[0]
            target_full = parts[1].split("(")[0]
            
            source_parts = source_full.split(".")
            target_parts = target_full.split(".")
            
            # Expected format: PImplementation.MachineName.StateName
            if len(source_parts) >= 3 and len(target_parts) >= 3:
                machine_name = source_parts[1]
                source_state = source_parts[2]
                target_state = target_parts[2]
                
                # Check if target belongs to the same machine
                if machine_name not in machines:
                    machines[machine_name] = []
                
                machines[machine_name].append(f"        {source_state} --> {target_state}: {val}")

    for machine, transitions in machines.items():
        mermaid.append(f"    state {machine} {{")
        mermaid.extend(list(sorted(set(transitions))))
        mermaid.append("    }")

    return "\n".join(mermaid)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python sci_to_mermaid.py <file.sci>")
        sys.exit(1)
    
    print(sci_to_mermaid(sys.argv[1]))
